proposals/interceptors-issues-2024-01.md

```diff
@@ -0,0 +1,147 @@
+
+```cs
+class C
+{
+    void M0() { }
+
+    void M()
+    {
+        Call();
+    }
+}
+```
+
+e.g. in the above, we might identify the invocation `Call` by saying:
+- it's in the first member of the compilation unit (`class C`)
+- it's in the second member of `class C`
+- it's in the first statement of method `M()`
+- it's in the expression of expression statement `Call();`
+- it's in the receiver of invocation expression `Call()`.
+
+This sequence could be encoded somehow and stuffed into the attribute. This would make it so that edits in other method bodies etc. will not necessarily require regenerating the interceptor. However, it's not clear how source generators, or the generator driver, could make use of this information in order to reduce the frequency of rerunning generators. Also, public API would likely be needed for this in order for generators to create the expected InterceptsLocationAttribute argument.
+
+### "Handle all"
+
+Some SGs might want to mitigate the need to rerun by simply "blanket handling" all calls to a certain set of methods in a certain context (e.g. an entire source file) at design time. Then during command line build we can actually generate granular calls.
+
+```cs
+public static void Usage()
+{
+    Original(42); // intercepted
+    Original(42, 43); // NOT intercepted, no diagnostic reported
+}
+
+public static void Original(int item) { }
+public static void Original(int item, int item2) { }
+
+[InterceptsLocation("path/to/file.cs", memberName: "Original")]
+public static void Interceptor(int item) { }
+```
+
+There is a risk with this approach that somehow a method would end up being intercepted with the "handle all" approach but not with the granular interceptors which are actually emitted.
+TODO: how to find misuse and help generator authors fix? maybe if *nothing* is intercepted by it?
+TODO: when we f12 on SG'd code, do we go to implementation code, even if SG is publishing "definition-only" versions of the sources?
+
+## Intercepting more member kinds
+
+Currently the interceptors feature only supports intercepting ordinary methods. Much like with partial methods and the upcoming partial properties, users want to be able to intercept usages of more kinds of members, such as properties and constructors.
+
+For all the member kinds we want to support, we will need to decide how usages of each member kind are denoted syntactically.
+
+For method invocations like `receiver.M(args)`, we use the location of the `M` token to intercept the call. Perhaps for property accesses like `receiver.P` we can use the `P` token. Things may get ambiguous if the property returns a delegate which is immediately invoked, e.g. `receiver.P()`. (TODO: do we support intercepting a delegate invocation? I think no, but check.)
+
+For constructors we may use either the `new` token or the type name. It's possible the type name would be more consistent in some sense, but it's not clear.
+
```

Comment:
The other item that needs to be decided is what should be passed for the `name` portion for constructors, property get / set, etc ... The good news is that there is prior art we can draw from here. The `UnsafeAccessor` feature has the same problems and has already come up with a naming scheme that we could re-use. This would likely be appreciated by customers as some, like EF, use both `UnsafeAccessor` and interceptors in the same generator. Similarity here would be appreciated. 


https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafeaccessorattribute?view=net-8.0